### DISCUSSION TITLE
RS485 VFD control over UART via TTL converter

### DISCUSSION THREAD

**Discussion by axg20202** [grblHAL/plugins#5] [Q&A]
Hi. I'm using Phil Barret's breakout board with a Teensy 4.1. I have a TTL to RS485 module hooked up to the serial pins (TX, RX, 3V3) on the BOB. Apprently my RS485 module can be run off 3.3 or 5v and has signal protection when using 5v, but I didn't want to risk it. Seems to power on with 3.3v. VFD (Huangyang) has both serial data connections going to the module in a twisted pair of cables.

IOsender sees the BOB no problem but I'm not able to get communication with the VFD . I get an ALARM14 at startup.  I have uncommented Modbus_enable and VFD_enable in my_machine.h.  I have also made all the changes on the VFD. I did wonder if the fact that I am connected via ethernet is the issue?  Can someone offer any suggestions please?

**Comment by phomann** [comment]
Hi,
Have a close look at the Huangyang manual.  It is my understanding that this brand of VFDs are not Modbus compliant. 
Do a search on Google. There were many discussions regarding getting these VFDs to communicate with Mach3. 
Cheers 

Peter Homann - (from my mobile)
http://www.homanndesigns.com

**Comment by terjeio** [comment]
There are [two Huanyang variants](https://github.com/grblHAL/iMXRT1062/blob/5660e14b39ebf7bb3235b90b7d1cc63e5543795d/grblHAL_Teensy4/src/my_machine.h#L58) implemented that are not compatible - are you using the correct one?

If there are LEDs on the modules are they showing any traffic?
Is the module auto direction capable or does it require a direction switching signal?

Are you using the latest version of grblHAL?

> It is my understanding that this brand of VFDs are not Modbus compliant.

Correct - at least for the oldest version which requires longer silent periods than specified.

**Reply by axg20202** [reply]
Hi terjeio, I've tried both variants of HY spindle implementation. My module does not require an additional direction switching signal. It has Tx, Rx, 3.3v and grnd, which i've connected to the corresponding pins on the BOB, which I understand to be serial1 of the teensy. When Iosender is running, there is flickering Tx LED activity on the module, which I believe means that it is polling the VFD for information, but not sure if this is actually what is happening. The Rx LED does not light. The VFD has been set to receive on the communication port and at the correct baud rate. I have tried switching polarity of signal wires. I've ordered a different RS485 board to try out as I feel I may have exhausted other avenues. However, I did wonder if the pin mapping for the BOB needs to be adjusted to get the serial port working properly?

**Reply by terjeio** [reply]
I have now wired my CNC simulator to check (I do not have a VFD to test with) and it is working for me.
Since the TX light is blinking data is sent - what you are missing is the response from the VFD so this means that either the data is either completely lost or that the VFD does not understand the messages.

Which model of VFD do you have? And again, which grblHAL version?

**Reply by axg20202** [reply]
I'm using XL sender edge. VFD is HY02D223B.

**Reply by terjeio** [reply]
Ok, it seems the VFD is using the [non-compliant ModBus message format](https://www.cnczone.com/forums/spindles-vfd/269196-rs485-commands-huanyang-vfd-hy02d223b.html) - so `#define VFD_ENABLE          1` is likely the correct configuration.

You can find the grblHAL version via _Help > About_ in the sender:

![image](https://user-images.githubusercontent.com/20260062/158804508-3721d18a-efec-4240-869e-5cf660921689.png)

**Reply by axg20202** [reply]
Yep, I tried both but currently set at 1.   I also discovered an undocumented setting for PD002 (source of operating frequency) on the HY VFD settings. Setting this to 2 is supposed to enable freq setting via RS485. If I set it to 3 (no description of this setting in the manual), I can get the spindle to turn on and off in either CW or CCW direction from Iosender and it reports the speed in Iosender too! However, it doesn't appear to spin up to the set frequency - seems like it spins to max speed regardless. Very strange, but does show that the BOB and RS485 board can send commands to the VFD and cause an effect.

**Reply by axg20202** [reply]
![image](https://user-images.githubusercontent.com/58083169/158806236-7709c268-25e2-4bf2-9c33-62d9a32044ca.png)

**Reply by axg20202** [reply]
Well, I don't think I changed anything at all, and all of a sudden it is working properly! All I did was change to the above undocumented setting, power cycled, set it back to 2 (RS485), power cycled again, and now everything works. Very strange indeed.
